SYCL/ESIMD/api/esimd_merge.cpp

```diff
@@ -0,0 +1,123 @@
+//==---------------- esimd_merge.cpp  - DPC++ ESIMD on-device test ---------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// REQUIRES: gpu
+// UNSUPPORTED: cuda || hip
+// RUN: %clangxx -fsycl %s -o %t.out
+// RUN: %HOST_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+
+// This is a functional test for esimd::merge free functions. 
+
+#include "../esimd_test_utils.hpp"
+
+#include <CL/sycl.hpp>
+#include <iostream>
+#include <sycl/ext/intel/experimental/esimd.hpp>
+
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace sycl::ext::intel::experimental;
+using namespace cl::sycl;
+
+template <class T> void prn(T *arr, int size, const char *title) {
+  std::cout << title << ": ";
+  for (int i = 0; i < size; ++i) {
+    std::cout << " " << arr[i];
+  }
+  std::cout << "\n";
+}
+
+int main(void) {
+  constexpr unsigned VL = 16;
+  constexpr unsigned FACTOR = 2;
+  constexpr unsigned SUB_VL = VL / FACTOR / FACTOR;
+  constexpr unsigned Size = VL * 2;
+
+  queue q(esimd_test::ESIMDSelector{}, esimd_test::createExceptionHandler());
+
+  auto dev = q.get_device();
+  std::cout << "Running on " << dev.get_info<info::device::name>() << "\n";
+
+  int *A = malloc_shared<int>(Size, q);
+  int *B = malloc_shared<int>(Size, q);
+  using MaskElT = typename simd_mask<1>::element_type;
+  MaskElT *M = malloc_shared<MaskElT>(Size, q);
+  int *C = malloc_shared<int>(Size, q);
+  constexpr int VAL0 = 0;
+  constexpr int VAL1 = 1;
+  constexpr int VAL2 = 3;
+
+  for (int i = 0; i < Size; ++i) {
+    A[i] = i % 2 + VAL1; // 1212 ...
+    B[i] = i % 2 + VAL2; // 3434 ...
+    // mask out first half of sub-vector, alternating '1' and '2' as 'enabled'
+    // bit representation in a mask element:
+    M[i] = (i % SUB_VL) >= (SUB_VL / 2) ? (i % SUB_VL - 1) : 0; // 00120012 ...
+    C[i] = VAL0;
+  }
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for<class Test>(Size, [=](id<1> i) SYCL_ESIMD_KERNEL {
+        simd<int, VL> va(A + i * VL);
+        simd<int, VL> vb(B + i * VL);
+        simd_mask<SUB_VL> m(M + i * VL);
+        // va:         1212121212121212
+        // vb:         3434343434343434
+        // m:          0012001200120012
+        // va.sel.sel: 1111
+        // vb.sel.sel: 4444
+        // vc:         1144
+        simd<int, SUB_VL> vc = esimd::merge(
+            va.template select<SUB_VL * 2, 2>(0).template select<SUB_VL, 1>(1),
```

Comment:
Minor comment. Because the 'va' is not parametrized by function template args it should be possible to omit the ".template " usage here.